File: server/camcops_server/tasks/ctqsf.py
#!/usr/bin/env python

"""
camcops_server/tasks/ctqsf.py

===============================================================================

    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.

===============================================================================

**Childhood Trauma Questionnaire, Short Form (CTQ-SF).**

"""

from cardinal_pythonlib.classes import classproperty
from cardinal_pythonlib.stringfunc import strseq
from semantic_version import Version

from camcops_server.cc_modules.cc_request import CamcopsRequest
from camcops_server.cc_modules.cc_task import Task, TaskHasPatientMixin


# =============================================================================
# CTQ-SF
# =============================================================================


class Ctqsf(TaskHasPatientMixin, Task):
    """
    Server implementation of the CTQ-SF task.
    """

    __tablename__ = "ctqsf"
    shortname = "CTQ-SF"
    provides_trackers = False

    # todo: Ctqsf fields

    N_QUESTIONS = 28
    QUESTION_FIELDNAMES = strseq("q", 1, N_QUESTIONS)

    @staticmethod
    def longname(req: "CamcopsRequest") -> str:
        _ = req.gettext
        return _("Childhood Trauma Questionnaire, Short Form")

    # noinspection PyMethodParameters
    @classproperty
    def minimum_client_version(cls) -> Version:
        return Version("2.2.8")

    def is_complete(self) -> bool:
        return self.all_fields_not_none(self.QUESTION_FIELDNAMES)

    def get_task_html(self, req: CamcopsRequest) -> str:
        return ""  # todo: IMPLEMENT Ctqsf

    # No SNOMED code for the CTQ.

File: tablet_qt/tasks/ctqsf.cpp
/*
    Copyright (C) 2012, University of Cambridge, Department of Psychiatry.
    Created by Rudolf Cardinal (rnc1001@cam.ac.uk).

    This file is part of CamCOPS.

    CamCOPS is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    CamCOPS is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with CamCOPS. If not, see <https://www.gnu.org/licenses/>.
*/

#if 0

    #include "ctqsf.h"

    #include "lib/stringfunc.h"
    #include "lib/version.h"
    #include "tasklib/taskfactory.h"
    #include "tasklib/taskregistrar.h"
using stringfunc::strseq;

const int FIRST_Q = 1;
const int N_QUESTIONS = 28;
const QString QPREFIX("q");

const QString Ctqsf::CTQSF_TABLENAME("ctqsf");


void initializeCtqsf(TaskFactory& factory)
{
    static TaskRegistrar<Ctqsf> registered(factory);
}


Ctqsf::Ctqsf(CamcopsApp& app, DatabaseManager& db, const int load_pk) :
    Task(app, db, CTQSF_TABLENAME, false, false, false)
        // ... anon, clin, resp
{
    addFields(strseq(QPREFIX, FIRST_Q, N_QUESTIONS), QMetaType::fromType<int>());

    load(load_pk);  // MUST ALWAYS CALL from derived Task constructor.
}


// ============================================================================
// Class info
// ============================================================================

QString Ctqsf::shortname() const
{
    return "CTQ-SF";
}


QString Ctqsf::longname() const
{
    return tr("Childhood Trauma Questionnaire, Short Form");
}


QString Ctqsf::description() const
{
    return tr("28-item self-rating scale of adverse childhood experiences.");
}


Version Ctqsf::minimumServerVersion() const
{
    return Version(2, 3, 3);
}


// ============================================================================
// Instance info
// ============================================================================

bool Ctqsf::isComplete() const
{
    return !anyValuesNull(strseq(QPREFIX, FIRST_Q, N_QUESTIONS));
}


QStringList Ctqsf::summary() const
{
    return QStringList{
        // ***
    };
}


QStringList Ctqsf::detail() const
{
    const QString spacer = " ";
    QStringList lines = completenessInfo();
    // *** lines
    lines.append("");
    lines += summary();
    return lines;
}


OpenableWidget* Ctqsf::editor(const bool read_only)
{
    // ***
    return nullptr;
}


// ============================================================================
// Task-specific calculations
// ============================================================================

#endif
